### PR TITLE
importccl: Support mysql's OUTFILE format in IMPORT

### PR DESCRIPTION
**Pull request by dt**
This adds support for MySQL's OUTFILE (as created by `mysqldump --tab`) to IMPORT.

### CODE REVIEW

File: pkg/ccl/importccl/read_import_mysqlout.go [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+			}
+
+			if c == d.opts.Escape {
+				nextLiteral = true
+				continue
+			}
+		}
+
+		// If we have a defined enclose char, check if we're starting or stopping
+		// an enclosed field. Technically, the enclose char can be made mandatory,
```
Comment (mjibson) [inline pkg/ccl/importccl/read_import_mysqlout.go]: Is the flag described in this comment d.opts.Enclose? If so fix this comment.

File: pkg/ccl/importccl/read_import_mysqlout.go [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+		// so potentially we could have a flag to enforce that, returning an error
+		// if it is missing, but we don't need to care for simply decoding.
+		if d.opts.Enclose != roachpb.MySQLOutfileOptions_Never && c == d.opts.Encloser {
+			readingField = !readingField
```
Comment (mjibson) [inline pkg/ccl/importccl/read_import_mysqlout.go]: if d.opts.Encloser appears multiple times in a row they will just keep flipping readingField. Is that intended and correct?

File: pkg/ccl/importccl/read_import_mysqlout.go [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+		}
+
+		// Are we done with the field, or even the whole row?
+		if !readingField && (c == d.opts.FieldSeparator || c == d.opts.RowSeparator) {
+			row = append(row, string(field))
+			field = field[:0]
+
+			if c == d.opts.RowSeparator {
+				if expected := d.csvInputReader.expectedCols; expected != len(row) {
+					return errors.Errorf("row %d: expected %d columns, go %d: %#v", count+1, expected, len(row), row)
```
Comment (mjibson) [inline pkg/ccl/importccl/read_import_mysqlout.go]: there's a nice error message here, but not in any other error return path. i think we should make a general solution (like a struct with an Error method) to not introduce more things as described in #25532. doesn't need to be done in this pr.

File: pkg/ccl/importccl/read_import_mysqlout_test.go [hunk 4]
```diff
@@ -0,0 +1,312 @@
+// Copyright 2018 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (mjibson) [inline pkg/ccl/importccl/read_import_mysqlout_test.go]: Should this be CCL?

File: pkg/ccl/importccl/read_import_mysqlout_test.go [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+		r.Seed(int64(i))
+		r.Read(buf)
+		testRows = append(testRows, testRow{i: i + 100, s: randStr(r, badChars, 1000), b: buf})
+	}
+
+	configs := []dumpCfg{
+		{
+			name: "escape-and-enclose",
+			opts: roachpb.MySQLOutfileOptions{
+				FieldSeparator: '\t', RowSeparator: '\n', HasEscape: true, Escape: '\\', Enclose: roachpb.MySQLOutfileOptions_Always, Encloser: '"',
```
Comment (mjibson) [inline pkg/ccl/importccl/read_import_mysqlout_test.go]: these would be easier to read separated by newlines

File: pkg/ccl/importccl/read_import_mysqlout_test.go [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+}
+
+func TestMysqlOutfileReader(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	testRows, configs := getMysqlOutfileTestdata(t)
+	ctx := context.TODO()
+	for _, config := range configs {
+		t.Run(config.name, func(t *testing.T) {
+			converter := newMysqloutfileReader(ctx, nil, config.opts, nil, 3)
```
Comment (mjibson) [inline pkg/ccl/importccl/read_import_mysqlout_test.go]: could this 3 be a named const?

File: pkg/ccl/importccl/read_import_mysqlout_test.go [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+			for b := range converter.csvInputReader.recordCh {
+				batches++
+				for _, r := range b.r {
+					if len(r) != 3 {
+						t.Fatalf("bad row len for %v", r)
+					}
+					// If the csv writing part of our conversion encodes as hex, we can
+					// round-trip bytes though the go csv parser.
+					orig := r[2]
+					r[2] = hex.EncodeToString([]byte(r[2]))
```
Comment (mjibson) [inline pkg/ccl/importccl/read_import_mysqlout_test.go]: i don't understand what's going on here. why don't we have to append `\x` to this?

File: pkg/ccl/importccl/read_import_mysqlout_test.go [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+				}
+			}
+
+			for i, row := range testRows {
+				expected := row.b
+				if expected == nil {
+					expected = []byte(config.null)
+				}
+
+				actual := []byte(res[i][2])
```
Comment (mjibson) [inline pkg/ccl/importccl/read_import_mysqlout_test.go]: i don't see where res[i][0] is verified

File: pkg/util/strings.go [hunk 9]
```diff
@@ -25,12 +25,12 @@ func GetSingleRune(s string) (rune, error) {
...
 		return 0, nil
 	}
-	r, sz := utf8.DecodeRuneInString(s)
+	r, _ := utf8.DecodeRuneInString(s)
 	if r == utf8.RuneError {
 		return 0, errors.Errorf("invalid character: %s", s)
 	}
-	if sz != len(s) {
-		return r, errors.New("must be only one character")
+	if l := len(s); 1 != l {
```
Comment (mjibson) [inline pkg/util/strings.go]: I think it was correct before. Add a descriptive comment if this is indeed intended. The length of a string is the number of bytes, not the number of runes. See https://play.golang.org/p/SE0ybC7jWoN

File: pkg/ccl/importccl/read_import_mysqlout_test.go [hunk 10]
```diff
@@ -0,0 +1,312 @@
...
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"testing"
+	"unicode/utf8"
+
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+	_ "github.com/go-sql-driver/mysql"
```
[the remark this answers is not in the document]
Comment (dt) [reply pkg/ccl/importccl/read_import_mysqlout_test.go]: We already vendor it: https://github.com/cockroachdb/cockroach/blob/master/Gopkg.lock#L413

**Comment by dt** [comment]
Review status: 0 of 21 files reviewed at latest revision, 14 unresolved discussions.

---

*[pkg/ccl/importccl/import_stmt.go, line 393 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25615#-LCid--6-bCWQdZ46ChJ-r3-393:-LCka_JN6azwgqrOIkbe:bkwkdsl) ([raw file](https://github.com/cockroachdb/cockroach/blob/e88994b6f80d5ef8396a91a804c1afc153a43295/pkg/ccl/importccl/import_stmt.go#L393)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

Is it better to move these checks to be per format instead of anything not CSV? When we add mysql SQL support we'll need another version bump and we'll have to split it apart there again.
</blockquote></details>

I think we just need to check for not CSV: if all the nodes are aware of the IOFormat proto, they'll know to look at the enum, and correctly fail if they don't support a format. I added a better error for that in the processor.

---

*[pkg/ccl/importccl/import_stmt_test.go, line 1181 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25615#-LCid--6-bCWQdZ46ChL-r3-1181:-LClAbCk5GIVMlYcTy3z:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e88994b6f80d5ef8396a91a804c1afc153a43295/pkg/ccl/importccl/import_stmt_test.go#L1181)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

spelling? `rowi`
</blockquote></details>

Done.

---

*[pkg/ccl/importccl/read_import_mysqlout.go, line 47 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25615#-LCid--7Ev59mXJIinfG-r2-47:-LClNJvY5JFFFamE8cSf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c3fa62ff3ac20f1f4ae94029edfcfe947fe7ac2/pkg/ccl/importccl/read_import_mysqlout.go#L47)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

this type is used a bunch now and maybe should get its own type? don't care too much.
</blockquote></details>

Done.

---

*[pkg/ccl/importccl/read_import_mysqlout.go, line 147 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25615#-LCid--7Ev59mXJIinfG-r2-147:-LClX4eg3Wha8IUdE9bP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c3fa62ff3ac20f1f4ae94029edfcfe947fe7ac2/pkg/ccl/importccl/read_import_mysqlout.go#L147)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

Is the flag described in this comment d.opts.Enclose? If so fix this comment.
</blockquote></details>

Done.

---

*[pkg/ccl/importccl/read_import_mysqlout.go, line 151 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25615#-LCid--7Ev59mXJIinfG-r2-151:-LClNNX23SBAgdrdJfQt:b-bvyxui) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c3fa62ff3ac20f1f4ae94029edfcfe947fe7ac2/pkg/ccl/importccl/read_import_mysqlout.go#L151)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

if d.opts.Encloser appears multiple times in a row they will just keep flipping readingField. Is that intended and correct?
</blockquote></details>

Eh, that's more or less my read of the docs. They say it has to be escaped if it occurs in the content.

---

*[pkg/ccl/importccl/read_import_mysqlout.go, line 162 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25615#-LCid--7Ev59mXJIinfG-r2-162:-LClO875AfcmJfRaCXEJ:b-d86wkc) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c3fa62ff3ac20f1f4ae94029edfcfe947fe7ac2/pkg/ccl/importccl/read_import_mysqlout.go#L162)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

there's a nice error message here, but not in any other error return path. i think we should make a general solution (like a struct with an Error method) to not introduce more things as described in #25532. doesn't need to be done in this pr.
</blockquote></details>

hm, there aren't actually many error returns mid-file -- I think it is just this one and then io errors from reading bytes from the reader. There are two error returns that can happen at end-of-file.

---

*[pkg/ccl/importccl/read_import_mysqlout_test.go, line 3 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25615#-LCid--7Ev59mXJIinfH-r2-3:-LClP0GY0H-C03NVM2Bp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c3fa62ff3ac20f1f4ae94029edfcfe947fe7ac2/pkg/ccl/importccl/read_import_mysqlout_test.go#L3)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

Should this be CCL?
</blockquote></details>

Done.

---

*[pkg/ccl/importccl/read_import_mysqlout_test.go, line 170 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25615#-LCid--7Ev59mXJIinfH-r2-170:-LClP0h_1RdqnklUrF4e:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c3fa62ff3ac20f1f4ae94029edfcfe947fe7ac2/pkg/ccl/importccl/read_import_mysqlout_test.go#L170)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

these would be easier to read separated by newlines
</blockquote></details>

Done.

---

*[pkg/ccl/importccl/read_import_mysqlout_test.go, line 211 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25615#-LCid--7Ev59mXJIinfH-r2-211:-LClP8iTCUOVeFVQ6KZA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c3fa62ff3ac20f1f4ae94029edfcfe947fe7ac2/pkg/ccl/importccl/read_import_mysqlout_test.go#L211)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

could this 3 be a named const?
</blockquote></details>

Done.

---

*[pkg/ccl/importccl/read_import_mysqlout_test.go, line 295 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25615#-LCid--7Ev59mXJIinfH-r2-295:-LClPMsm2EvDFcICaPdE:b-g0gblp) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c3fa62ff3ac20f1f4ae94029edfcfe947fe7ac2/pkg/ccl/importccl/read_import_mysqlout_test.go#L295)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

i don't see where res[i][0] is verified
</blockquote></details>

res[i][0] wasn't explicitly verified, was just being used as a label in case the other cols didn't line up, but I guess I can verify it too.

---

*[pkg/ccl/importccl/read_import_proc.go, line 342 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25615#-LCid--7Ev59mXJIinfC-r2-342:-LClQPLTDpihPse3oPCM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c3fa62ff3ac20f1f4ae94029edfcfe947fe7ac2/pkg/ccl/importccl/read_import_proc.go#L342)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

all of these arguments should be named
</blockquote></details>

Done.

---

*[pkg/util/strings.go, line 32 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25615#-LCj3MVTyEWNtaCzF94q-r3-32:-LClSu5CCSA3gQJAzkv2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e88994b6f80d5ef8396a91a804c1afc153a43295/pkg/util/strings.go#L32)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

I think it was correct before. Add a descriptive comment if this is indeed intended. The length of a string is the number of bytes, not the number of runes. See https://play.golang.org/p/SE0ybC7jWoN
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25615)*

**Comment by mjibson** [comment]
Review status: 0 of 21 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/ccl/importccl/read_import_mysqlout.go, line 162 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25615#-LCid--7Ev59mXJIinfG-r2-162:-LCmArzn1lWvmaA8zBmG:b872wrt) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c3fa62ff3ac20f1f4ae94029edfcfe947fe7ac2/pkg/ccl/importccl/read_import_mysqlout.go#L162)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

hm, there aren't actually many error returns mid-file -- I think it is just this one and then io errors from reading bytes from the reader. There are two error returns that can happen at end-of-file.
</blockquote></details>

i'm specifically worried about the finished block above where it checks for expected enclosure chars and stuff. in any case can happen later.

---

*[pkg/ccl/importccl/read_import_mysqlout_test.go, line 34 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25615#-LCid--7Ev59mXJIinfH-r2-34:-LCmB3MgAqdR-yt9Q4WL:b2nhoia) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c3fa62ff3ac20f1f4ae94029edfcfe947fe7ac2/pkg/ccl/importccl/read_import_mysqlout_test.go#L34)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

We already vendor it: https://github.com/cockroachdb/cockroach/blob/master/Gopkg.lock#L413
</blockquote></details>

TIL

---

*[pkg/ccl/importccl/read_import_mysqlout_test.go, line 245 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25615#-LCid--7Ev59mXJIinfH-r2-245:-LCmBBdC4IEkRIN0RjMf:b-tdiqtw) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c3fa62ff3ac20f1f4ae94029edfcfe947fe7ac2/pkg/ccl/importccl/read_import_mysqlout_test.go#L245)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

i don't understand what's going on here. why don't we have to append `\x` to this?
</blockquote></details>

reminder about this question

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25615)*

**Comment by dt** [comment]
Review status: 0 of 21 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/ccl/importccl/read_import_mysqlout_test.go, line 245 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25615#-LCid--7Ev59mXJIinfH-r2-245:-LD2m-TW-Ern6CooP92I:bkehns) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c3fa62ff3ac20f1f4ae94029edfcfe947fe7ac2/pkg/ccl/importccl/read_import_mysqlout_test.go#L245)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

reminder about this question
</blockquote></details>

This is writing to a CSV, and later we read that CSV.

This was particularly useful when we were testing mysqlout->csv->cockroach. Now that we're doing mysqlout->cockroach, I'm not sure if we need to read and write csvs at all -- I kept them though since they're more human readable than the raw mysqlout.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25615)*

**Review by madelynnblue — APPROVED** [review]

**Comment by dt** [comment]
bors r=mjibson

**Pull request merged**